Small optimization in MHM build

klaussfreire opened the pull request:
Avoid some unnecessary work in _discard_duplicates when there
are no duplicates

klaussfreire committed:
Small optimization in MHM build

Avoid some unnecessary work in _discard_duplicates when there
are no duplicates

klaussfreire committed:
Bump version

klaussfreire committed:
Optimize tests too

They're annoyingly slow

klaussfreire committed:
More MHM optimizations

Apply the same optimization numpy.unique applies when sorting
multidimensional arrays: use a void type to do faster binary row
comparisons instead of structural comparisons.

While the sort order of those comparisons isn't adequate for the
final index, it works well as de-duplication comparison, so use
it for discard_duplicates unless it's one of the final
deduplication passes.

klaussfreire committed:
Remove unused variables

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2624,17 +2624,18 @@ def _merge_all(parts, dtype):
         elif discard_duplicates:
-            flags = numpy.concatenate([[True], vpart[1:,0] != vpart[:-1,0]])
+            flags = vpart[1:,0] != vpart[:-1,0]
         if not numpy.all(flags):
```
no need to call all now right?

klaussfreire replied:
Why not?

zeitos replied:
my bad, != is overriden

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2624,17 +2624,18 @@ def _merge_all(parts, dtype):
+            flags = vpart[1:,0] != vpart[:-1,0]
         if not numpy.all(flags):
+            flags = numpy.concatenate([[True], flags])
             apart = apart[flags]
```
the name is `discard_duplicates`
but:
```
array([[ 1.,  0.,  0.],
       [ 2.,  3.,  4.],
       [ 1.,  2.,  4.]])
>>> apart
array([[ 1.,  0.,  0.],
       [ 1.,  1.,  2.],
       [ 2.,  3.,  4.],
       [ 1.,  2.,  4.],
       [ 1.,  3.,  5.]])
>>> apart[1:,0] != apart[:-1,0]
array([False,  True,  True, False], dtype=bool)
>>> apart[[numpy.concatenate([[True],(apart[1:,0] != apart[:-1,0])])]]
array([[ 1.,  0.,  0.],
       [ 2.,  3.,  4.],
       [ 1.,  2.,  4.]])
```

klaussfreire replied:
You're using the `discard_duplicate_keys` code path, which is supposed to deduplicate based on the first column only. So it's correct.

Furthermore, you're skipping the sort pass, which makes the result also incorrect.

klaussfreire committed:
In-place array compression in discard_duplicates

Use an in-place compression operation to avoid copying large
arrays in discard_duplicates

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2634,9 +2635,33 @@ def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplic
+                count_nonzero = np.count_nonzero
+                start = nmin
+                end = len(flags)
+                out_start = start+1
```
why +1?

klaussfreire replied:
Because of the first implicit True

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2634,9 +2635,33 @@ def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplic
+                    flags_slice = flags[start:chunk_end]
+                    nout = count_nonzero(flags_slice)
+                    dedup_slice = apart[start+1:chunk_end+1]
+                    if nout != (chunk_end - start):
```
why not `nout < (chunk_end - start)`? it shouldn't be bigger

klaussfreire replied:
If it's bigger it's a bug, skipping the operation would silence that bug, carrying it out would make numpy raise an exception and expose it.

zeitos replied:
if that's the case wouldn't it be better to have our own exception that explains the reason?
```
else:
   if nout > (chunk_end - start):
       raise ValueError(("Since python can't create matter, "
            "there's got to be something really wrong. We found "
            "more non zeros than the array size! some context info:"
            " end: %s, start: %s, nout: %s, slice: %s"), end, start, nout, flags_slice)
```

klaussfreire replied:
No, numpy's exception is very precise and informative

klaussfreire replied:
I'm going to cry timeout here...

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2634,9 +2635,33 @@ def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplic
+                    chunk_end = min(end, start + 100000)
+                    flags_slice = flags[start:chunk_end]
+                    nout = count_nonzero(flags_slice)
+                    dedup_slice = apart[start+1:chunk_end+1]
```
I don't understand why the +1s here :S

klaussfreire replied:
Because the flags array is missing the initial True, this compensates for that implicit True

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2624,18 +2624,44 @@ def _merge_all(parts, dtype):
+                # result into the input array directly.
+                nmin = np.argmin(flags)
+                count_nonzero = np.count_nonzero
+                start = nmin
```
Why can't you do `start = nmin + 1` to prevent doing the `+1` all the time?

klaussfreire replied:
If you read carefully, you'll see that the start/end/chunk_end variables get used in both contexts, those that need the +1 and those that don't.

klaussfreire committed:
Consider nout=0 case
